File: src/main/java/ru/pshiblo/services/audio/local/LocalMusicService.java
package ru.pshiblo.services.audio.local;

import com.sedmelluq.discord.lavaplayer.format.AudioDataFormat;
import com.sedmelluq.discord.lavaplayer.format.AudioPlayerInputStream;
import com.sedmelluq.discord.lavaplayer.player.AudioPlayer;
import com.sedmelluq.discord.lavaplayer.player.AudioPlayerManager;
import com.sedmelluq.discord.lavaplayer.player.DefaultAudioPlayerManager;
import com.sedmelluq.discord.lavaplayer.source.AudioSourceManagers;
import com.sedmelluq.discord.lavaplayer.track.AudioTrack;
import ru.pshiblo.services.MusicService;
import ru.pshiblo.services.ServiceThread;
import ru.pshiblo.gui.log.ConsoleOut;
import ru.pshiblo.services.ServiceType;
import ru.pshiblo.services.audio.AudioFactory;
import ru.pshiblo.services.audio.AudioLoadHandler;
import ru.pshiblo.services.audio.TrackScheduler;

import javax.sound.sampled.*;
import java.io.IOException;
import java.util.ArrayDeque;
import java.util.Queue;

import static com.sedmelluq.discord.lavaplayer.format.StandardAudioDataFormats.COMMON_PCM_S16_BE;

public class LocalMusicService extends ServiceThread implements MusicService {

    private final AudioPlayer player;
    private final TrackScheduler scheduler;

    public LocalMusicService() {
        AudioFactory.getInstance().setLocalConfig();
        player = AudioFactory.getInstance().createAudioPlayer();
        scheduler = new TrackScheduler(player);
        player.addListener(scheduler);
    }

    @Override
    protected void runInThread() {
        try {

            AudioDataFormat format = AudioFactory.getInstance().getPlayerManager().getConfiguration().getOutputFormat();
            AudioInputStream stream = AudioPlayerInputStream.createStream(player, format, 10000L, false);
            SourceDataLine.Info info = new DataLine.Info(SourceDataLine.class, stream.getFormat());
            SourceDataLine line = (SourceDataLine) AudioSystem.getLine(info);
            line.open(stream.getFormat());
            line.start();

            byte[] buffer = new byte[COMMON_PCM_S16_BE.maximumChunkSize()];
            int chunkSize;

            while ((chunkSize = stream.read(buffer)) >= 0) {
                if (!scheduler.queueIsEmpty()) {
                    line.write(buffer, 0, chunkSize);
                }
            }
        }catch (LineUnavailableException | IOException e) {
            e.printStackTrace();
        }
    }

    @Override
    public void play(String track) {
        AudioFactory.getInstance().loadItemOrdered(player, track, new AudioLoadHandler(track, scheduler));
    }

    @Override
    public void skip() {
        scheduler.skipTrack();
    }

    @Override
    public void volume(int volume) {
        player.setVolume(volume);
    }

    @Override
    public String getPlayingTrack() {
        AudioTrack playingTrack = player.getPlayingTrack();

        if (playingTrack == null)
            return null;

        return playingTrack.getInfo().title;
    }

    @Override
    public ServiceType getServiceType() {
        return ServiceType.MUSIC;
    }

    @Override
    public void shutdown() {
        player.destroy();
        super.shutdown();
    }
}

File: src/jvmMain/kotlin/ru/pshiblo/luna/services/audio/local/LocalMusicService.kt
package ru.pshiblo.luna.services.audio.local

import com.google.inject.Inject
import com.google.inject.Singleton
import com.sedmelluq.discord.lavaplayer.format.AudioPlayerInputStream
import com.sedmelluq.discord.lavaplayer.format.StandardAudioDataFormats
import kotlinx.coroutines.yield
import ru.pshiblo.luna.services.ServiceThread
import ru.pshiblo.luna.services.audio.MusicManager
import ru.pshiblo.luna.services.audio.MusicService
import ru.pshiblo.luna.services.toast.ToastHolderService
import java.io.IOException
import javax.sound.sampled.*

@Singleton
class LocalMusicService : ServiceThread(), MusicService {

    @Inject
    private lateinit var musicManager: MusicManager
    @Inject
    private lateinit var toastHolderService: ToastHolderService

    override suspend fun runInThread() {
        try {
            musicManager.setLocalConfig()
            toastHolderService.showMessage("Локальная музыка инициализирована")

            val format = musicManager.musicManager.configuration.outputFormat
            val stream = AudioPlayerInputStream.createStream(
                musicManager.musicPlayer,
                format,
                10000L,
                true
            )
            val info = DataLine.Info(SourceDataLine::class.java, stream.format)
            val line = AudioSystem.getLine(info) as SourceDataLine
            line.open(stream.format)
            line.start()
            val buffer = ByteArray(StandardAudioDataFormats.COMMON_PCM_S16_BE.maximumChunkSize())
            var chunkSize: Int
            while (stream.read(buffer).also { chunkSize = it } >= 0) {
                yield()
                if (!musicManager.trackScheduler.queueIsEmpty()) {
                    line.write(buffer, 0, chunkSize)
                }
            }
        } catch (e: LineUnavailableException) {
            e.printStackTrace()
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }

    override fun play(track: String) {
        if (isInit)
            musicManager.loadTrack(track)
    }

    override fun skip() {
        if (isInit)
            musicManager.trackScheduler.skipTrack()
    }

    override var volume: Int
        get() {
            return musicManager.musicPlayer.volume
        }
        set(value) {
            if (isInit)
                musicManager.musicPlayer.volume = value
        }

    override val playingTrack: String?
        get() = musicManager.musicPlayer.playingTrack?.info?.title
}